src/test/msgr/test_async_networkstack.cc

```diff
@@ -0,0 +1,1050 @@
+  }
+  EventCenter *get_center(unsigned i) {
+    return &stack->get_worker(i)->center;
+  }
+  Worker *get_worker(unsigned i) {
+    return stack->get_worker(i);
+  }
+  template<typename func>
+  class C_dispatch : public EventCallback {
+    Worker *worker;
+    func f;
+    bool done;
+   public:
+    C_dispatch(Worker *w, func &&_f): worker(w), f(std::move(_f)), done(false) {}
+    void do_request(int id) {
+      f(worker);
+      done = true;
+    }
+    void wait() {
+      int us = 1000 * 1000 * 1000;
+      while (!done) {
+        ASSERT_TRUE(us > 0);
+        usleep(100);
+        us -= 100;
+      }
+    }
+  };
+  template<typename func>
+  void exec_events(func &&f) {
+    std::vector<C_dispatch<func>*> dis;
+    for (unsigned i = 0; i < stack->get_num_worker(); ++i) {
+      Worker *w = stack->get_worker(i);
+      C_dispatch<func> *e = new C_dispatch<func>(w, std::move(f));
+      stack->get_worker(i)->center.dispatch_event_external(e);
+      dis.push_back(e);
+    }
+
+    for (auto &&e : dis) {
+      e->wait();
+      delete e;
+    }
+  }
+};
+
+class C_poll : public EventCallback {
+  EventCenter *center;
+  std::atomic<bool> wakeuped;
+  static const int sleepus = 500;
+  Mutex *lock;
+
+ public:
+  C_poll(EventCenter *c, Mutex *l=nullptr): center(c), wakeuped(false), lock(l) {}
+  void do_request(int r) {
+    wakeuped = true;
+  }
+  bool poll(int milliseconds) {
+    auto start = ceph::coarse_real_clock::now(g_ceph_context);
+    while (!wakeuped) {
+      if (lock)
+        lock->Lock();
+      center->process_events(sleepus);
+      if (lock)
+        lock->Unlock();
+      usleep(sleepus);
+      auto r = std::chrono::duration_cast<std::chrono::milliseconds>(
+              ceph::coarse_real_clock::now(g_ceph_context) - start);
+      if (r >= std::chrono::milliseconds(milliseconds))
+        break;
+    }
+    return wakeuped;
+  }
+  void reset() {
+    wakeuped = false;
+  }
+};
+
+TEST_P(NetworkWorkerTest, SimpleTest) {
+  entity_addr_t bind_addr;
+  ASSERT_EQ(bind_addr.parse(get_addr().c_str()), true);
+  std::atomic_bool accepted(false);
+  std::atomic_bool *accepted_p = &accepted;
+
+  exec_events([this, accepted_p, bind_addr](Worker *worker) mutable {
+    entity_addr_t cli_addr;
+    SocketOptions options;
+    ServerSocket bind_socket;
+    EventCenter *center = &worker->center;
+    ssize_t r = 0;
+    if (stack->support_local_listen_table() || worker->id == 0)
+      r = worker->listen(bind_addr, options, &bind_socket);
+    ASSERT_EQ(r, 0);
+
+    ConnectedSocket cli_socket, srv_socket;
+    if (worker->id == 0) {
+      r = worker->connect(bind_addr, options, &cli_socket);
+      ASSERT_EQ(r, 0);
+    }
+
+    bool is_my_accept = false;
+    if (bind_socket) {
+      C_poll cb(center);
+      center->create_file_event(bind_socket.fd(), EVENT_READABLE, &cb);
+      if (cb.poll(500)) {
+        *accepted_p = true;
+        is_my_accept = true;
+      }
+      ASSERT_EQ(*accepted_p, true);
```

Comment:
better off using `ASSERT_TRUE()` or use `ASSERT_EQ(true, *accepted_p)`. the first parameter is the expected value, and the second one is the actual result.